src/dbnode/storage/flush.go

```diff
@@ -147,42 +191,41 @@ func (m *flushManager) Flush(
 		multiErr = multiErr.Add(err)
 	}
 
-	err = m.rotateCommitlogAndSnapshot(namespaces, tickStart)
-	if err != nil {
-		multiErr = multiErr.Add(err)
-	}
+	return multiErr.FinalError()
+}
 
-	indexFlush, err := m.pm.StartIndexPersist()
+func (m *flushManager) dataColdFlush(
+	namespaces []databaseNamespace,
+	tickStart time.Time,
+) error {
+	flushPersist, err := m.pm.StartFlushPersist()
 	if err != nil {
-		multiErr = multiErr.Add(err)
-		return multiErr.FinalError()
+		return err
 	}
 
-	m.setState(flushManagerIndexFlushInProgress)
+	m.setState(flushManagerColdFlushInProgress)
+	multiErr := xerrors.NewMultiError()
 	for _, ns := range namespaces {
-		var (
-			indexOpts    = ns.Options().IndexOptions()
-			indexEnabled = indexOpts.Enabled()
-		)
-		if !indexEnabled {
-			continue
+		rOpts := ns.Options().RetentionOptions()
+		latestWarmFlushTime := retention.FlushTimeEnd(rOpts, tickStart)
```

Comment:
This definitely deserves a comment haha.

```
The current assumption of the warm flush system is that every warm flush will be the first time data for that namespace/shard/blockStart is written to disk and that no other files should be present. As a result, cold flushes for blocks that have not been warm flushed yet need to be avoided.

We avoid cold flushing blocks that have not yet been warm flushed by:

1. Performing warm flushes before cold flushes and not allowing cold flushes to proceed if the warm flushes failed.
2. Only performing flushes up to the latest block that is warm flushable to prevent cold flushes from flushing blocks that have not yet been made available to warm flush.
```

Reply:
@justinjc I gues per our discussion about moving this check into the ColdFlush method itself such that it will never attempt to ColdFlush a block that hasn't been WarmFlushed yet, we won't need this.

Reply:
👍